### PR TITLE
[Barcode Scanning] Excessive resource usage in barcodescanner

### LINKED ISSUE #10303: [Barcode Scanning] Excessive resource usage in `BarcodeScanner`
It looks like that we create flow for every single frame that comes from `ImageAnalysis` and we launch a new coroutine and collect (suspend). And all that on the main thread. I am surprised that we don't have any issues with that

We probably should move all the frame analyses from the compose view, and have 1 flow of frames, that is mapped with the bar code recognition function, and all empty frames should be filtered out

### PR DESCRIPTION
**Pull request by kidinov**
Closes: #10303 

### Description

As we found out in the current implementation, we create a flow on every frame that comes from the camera, and then we suspend it, which may be fine but still hard to follow. This PR refactors this approach a bit to keep it easier to follow and start/stop code recognition on demand

### Testing instructions

* Retest the flows where code scanner is used (Orders, Order creation)
* Notice no regrassions

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/barcodescanner/BarcodeScanner.kt [hunk 1]
```diff
@@ -51,34 +46,12 @@ fun BarcodeScanner(
...
-                )
-                    .setBackpressureStrategy(STRATEGY_KEEP_ONLY_LATEST)
-                    .build()
-                imageAnalysis.setAnalyzer(ContextCompat.getMainExecutor(context)) { imageProxy ->
-                    onScannedResult(codeScanner.startScan(imageProxy))
-                }
+                ).setBackpressureStrategy(STRATEGY_KEEP_ONLY_LATEST).build()
+                imageAnalysis.setAnalyzer(ContextCompat.getMainExecutor(context), onNewFrame)
                 try {
                     cameraProviderFuture.get().bindToLifecycle(lifecycleOwner, selector, preview, imageAnalysis)
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/barcodescanner/BarcodeScanner.kt]: @AnirudhBhat @samiuelson there is some magic happening. With the current implementation, after fragments pause -> resume, `imageAnalysis` stops receiving frames, while before for some mysterious reasons, that was not happening. My assumption is that it is related somehow to this code, which somehow causes recomposition

```
                        onScannedResult = { codeScannerStatus ->
                            viewLifecycleOwner.lifecycleScope.launch {
                                viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.CREATED) {
                                    codeScannerStatus.collect { status ->
                                        navigateBackWithResult(KEY_BARCODE_SCANNING_SCAN_STATUS, status)
                                    }
                                }
                            }
                        },
```
I am keep working on that

Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/barcodescanner/BarcodeScanner.kt]: Fixed that with restarting the camera and attaching the new instance of image analyzes on resume. Not clear for me why it does restart the preview internally but no analyzes 🤷‍♂️

3b65a046d80957d43b556249b3527976c765ba6a

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/barcodescanner/BarcodeScanningViewModel.kt [hunk 2]
```diff
@@ -2,24 +2,68 @@ package com.woocommerce.android.ui.barcodescanner
...
+    private var processingJob: Job? = null
+
     init {
         _permissionState.value = PermissionState.Unknown
     }
 
+    fun startCodesRecognition() {
+        frameChannel = createChannel()
+        processingJob = launch {
+            for (frame in frameChannel) {
```
Comment (samiuelson) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/barcodescanner/BarcodeScanningViewModel.kt:42]:
```suggestion
            for (frame in frameChannel) {
                if (!isActive) break
```
💡 We should probably check `isActive`. Otherwise having `processingJob` assigned to the property feels unnecessary. Wdyt?

I wonder if we should also add:
```kotlin 
override fun onCleared() {
        processingJob?.cancel()
    }
```
in case `stopCodesRecognition` is not called for some reason.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/barcodescanner/BarcodeScanningViewModel.kt [hunk 3]
```diff
@@ -62,6 +106,30 @@ class BarcodeScanningViewModel @Inject constructor(
...
+                CodeScannerStatus.Failure(
+                    error = exception.message,
+                    type = CodeScanningErrorType.Other(exception)
+                )
+            )
+        )
+        stopCodesRecognition()
+    }
+
+    private fun createChannel() = Channel<ImageProxy>(
```
Comment (samiuelson) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/barcodescanner/BarcodeScanningViewModel.kt:128]: ❤️ Neat idea to use the `Channel` here.

**Review by samiuelson — APPROVED** [review]
Thanks, @kidinov for your work! This is a significant improvement ❤️ 
 
I left a couple of minor questions, let me know what you think.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/barcodescanner/BarcodeScanningViewModel.kt (hunk 2 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/barcodescanner/BarcodeScanningViewModel.kt:42]: @samiuelson 

> 💡 We should probably check isActive. Otherwise having processingJob assigned to the property feels unnecessary. Wdyt?

Not sure I got this one. My understanding is that `for (frame in frameChannel) {` here we iterate over the channel and "hasNext" is a suspended function, and on every suspension point, there is a check if the job is active and if it, that exits, so manual check on the next line probably not necessary? 

```
override fun onCleared() {
        processingJob?.cancel()
    }
```

yep, good idea. Added here - cb2dd3b0b4f363bc1740588511f2caa1e72b251d

Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/barcodescanner/BarcodeScanningViewModel.kt:42]: > "hasNext" is a suspended function, and on every suspension point, there is a check if the job is active and if it, that exits, so manual check on the next line probably not necessary?

You're right 👍🏼

**Comment by kidinov** [comment]
@AnirudhBhat, I'd like you to take a look, as I may be missing something here. There is no rush (this is a hack week now) unless it is blocking for @samiuelson

**Comment by samiuelson** [comment]
> @AnirudhBhat, I'd like you to take a look, as I may be missing something here. There is no rush (this is a hack week now) unless it is blocking for @samiuelson

I'd like to ship "Scan to update inventory" this week. So if we could merge this PR soon it would be great 🙇🏼 cc @AnirudhBhat , @kidinov

**Review by AnirudhBhat — APPROVED** [review]
Great work here @kidinov Thank you for taking care of this. LGTM :shipit:
